types/query.d.ts

```diff
@@ -1,8 +1,23 @@
 declare module 'mongoose' {
   import mongodb = require('mongodb');
 
-  export type ApplyBasicQueryCasting<T> = T | T[] | (T extends (infer U)[] ? U : any) | any;
-  type Condition<T> = ApplyBasicQueryCasting<T> | QuerySelector<ApplyBasicQueryCasting<T>>;
+  type StringQueryTypeCasting = string | RegExp;
+  type ObjectIdQueryTypeCasting = Types.ObjectId | string;
+  type UUIDQueryTypeCasting = Types.UUID | string;
+  type BufferQueryTypeCasting = Buffer | any;
+
+  type QueryTypeCasting<T> = T extends string
+    ? StringQueryTypeCasting
+    : T extends Types.ObjectId
+      ? ObjectIdQueryTypeCasting
+      : T extends Types.UUID
+        ? UUIDQueryTypeCasting
+        : T extends Buffer
+          ? BufferQueryTypeCasting
+          : T;
+
+  export type ApplyBasicQueryCasting<T> = T | T[] | (T extends (infer U)[] ? QueryTypeCasting<U> : never);
```

Comment:
Does `ApplyBasicQueryCasting` still support arbitrary keys? Because it should.

Reply:
Chiming in here since I raised the original issue. Thanks for tackling @FaizBShah !

@vkarpov15 does `ApplyBasicQueryCasting` specifically need to accept arbitrary keys? Or is it just that `FilterQuery` should accept arbitrary keys? It appears that `RootQuerySelector` already accepts arbitrary keys, so we're covered on that front. Is there an example use case where there's a defined property of a schema where you need to query that specific property based on arbitrary value?

Reply:
That makes sense, I didn't look closely enough to see that `RootQuerySelector` supports arbitary keys. You're right that the important thing is for `FilterQuery` to support arbitrary keys. In that case this is fine.